torch/testing/_internal/common_quantized.py

```diff
@@ -43,20 +43,18 @@ def _calculate_dynamic_qparams(X, dtype, reduce_range=False):
     according to the min and max element of the tensor"""
     if isinstance(X, torch.Tensor):
         X = X.numpy()
-    if dtype == torch.qint8:
-        if reduce_range:
-            qmin, qmax = -64, 63
-        else:
-            qmin, qmax = -128, 127
-    else:  # dtype == torch.quint8
-        if reduce_range:
-            qmin, qmax = 0, 127
-        else:
-            qmin, qmax = 0, 255
-    n_levels = 255.0
+    if X.size == 0:
+        return float(1.0), int(0)
+    qinfo = torch.iinfo(dtype)
+    qmin = qinfo.min
+    qmax = qinfo.max
+    n_levels = float(qmax - qmin)
+    if reduce_range:
+        qmin = qmin // 2
+        qmax = qmax // 2
     min_val = X.min()
     max_val = X.max()
-    if min_val == max_val:
+    if min_val == max_val == 0:
```

Comment:
Dont you want to also check for min_val ==max_val also so that we avoid scale=0?

Reply:
I don't think we need it -- if `min_val == max_val` it will be caught in the later lines, when it says

```
max_val = max(max_val, 0.0)
min_val = min(min_val, 0.0)
```

Because we already checked if they are zero, we will get a valid range `[min_val, max_val]`, where they will be different, while only one of them might be equal to zero. That way, in case we have a tensor with all the same values in it, it will still be correctly representable, as we will have the range `[0, VALUE]`